### PR TITLE
[CONSRV] Preparation for Asian console support

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose
Make preparation for console support of Chinese, Japanese and Korean (CJK) codepages.
JIRA issue: [CORE-12451](https://jira.reactos.org/browse/CORE-12451)

- Add `consrv/frontends/wcwidth.c`.
- Add `GuiPaintTextModeBufferCJK` function.

### CODE REVIEW

**Comment by HBelusca** [comment]
So the standard CRT wcwidth function is incomplete for this job?

File: win32ss/user/winsrv/consrv/frontends/terminal.c [hunk 1]
```diff
@@ -491,6 +491,23 @@ ConioNextLine(PTEXTMODE_SCREEN_BUFFER Buff, PSMALL_RECT UpdateRect, PUINT Scroll
     UpdateRect->Bottom = Buff->CursorPosition.Y;
 }
 
+/* Chinese, Japanese or Korean? */
+BOOL FASTCALL
+IsCodePageCJK(UINT CodePage)
```
Comment (HBelusca) [inline win32ss/user/winsrv/consrv/frontends/terminal.c]: There exists already a macro for that: `IsCJKCodePage()` .

File: win32ss/user/winsrv/consrv/frontends/terminal.c [hunk 2]
```diff
@@ -505,6 +522,7 @@ ConioWriteConsole(PFRONTEND FrontEnd,
     SMALL_RECT UpdateRect;
     SHORT CursorStartX, CursorStartY;
     UINT ScrolledLines;
+    const BOOL bCJK = IsCodePageCJK(Console->OutputCodePage);
```
Comment (HBelusca) [inline win32ss/user/winsrv/consrv/frontends/terminal.c]: why const?

File: win32ss/user/winsrv/consrv/frontends/wcwidth.c [hunk 3]
```diff
@@ -0,0 +1,309 @@
+/*
```
Comment (HBelusca) [inline win32ss/user/winsrv/consrv/frontends/wcwidth.c]: Is the CRT wcwidth function incomplete? Or isn't there already another win32 function that does this job?

File: win32ss/user/winsrv/consrv/frontends/terminal.c (hunk 1 above)
Comment (katahiromz) [reply win32ss/user/winsrv/consrv/frontends/terminal.c]: That is defined in another module (`concfg`). Should we commonize?

File: win32ss/user/winsrv/consrv/frontends/terminal.c (hunk 2 above)
Comment (katahiromz) [reply win32ss/user/winsrv/consrv/frontends/terminal.c]: Removed.

File: win32ss/user/winsrv/consrv/frontends/wcwidth.c (hunk 3 above)
Comment (katahiromz) [reply win32ss/user/winsrv/consrv/frontends/wcwidth.c]: Some characters of the East Asian Ambiguous category are to be specially treated as fullwidth characters in Asian console. See the comments of `mk_wcwidth_cjk` function.

Comment (katahiromz) [reply win32ss/user/winsrv/consrv/frontends/wcwidth.c]: For example, Cyrillic Д is defined in Japanese Shift_JIS codepage, and this is a fullwidth character.

Comment (katahiromz) [reply win32ss/user/winsrv/consrv/frontends/wcwidth.c]: For more example, ｱ (`U+FF71 HALFWIDTH KATAKANA LETTER A`) is a halfwidth character.
```c
printf("%d, %d\n", mk_wcwidth(0xFF71), mk_wcwidth_cjk(0xFF71)); // 1, 1
```

Comment (katahiromz) [reply win32ss/user/winsrv/consrv/frontends/wcwidth.c]: We can use `WideCharToMultiByte(codepage, 0, &ch, 1, NULL, 0, NULL, NULL);` for getting width, but I think it's slow.

Comment (katahiromz) [reply win32ss/user/winsrv/consrv/frontends/wcwidth.c]:
```c
    DWORD dw1, dw2, dw3;
    WCHAR ch = 0x0414;

    DWORD dw1 = GetTickCount();
    for (int i = 0; i < 100000000; ++i)
        mk_wcwidth_cjk(0xFF71);
    DWORD dw2 = GetTickCount();
    for (int i = 0; i < 100000000; ++i)
        WideCharToMultiByte(932, 0, &ch, 1, NULL, 0, NULL, NULL);
    DWORD dw3 = GetTickCount();

    printf("%d, %d\n", dw2 - dw1, dw3 - dw2);
```

無題
`WideCharToMultiByte` is faster...?

Comment (katahiromz) [reply win32ss/user/winsrv/consrv/frontends/wcwidth.c]: `wcwidth.c` is deleted.

Comment (katahiromz) [reply win32ss/user/winsrv/consrv/frontends/wcwidth.c]: No, it was not optimized.
無題

Comment (katahiromz) [reply win32ss/user/winsrv/consrv/frontends/wcwidth.c]: `wcwidth.c` is added.

Comment (katahiromz) [reply win32ss/user/winsrv/consrv/frontends/wcwidth.c]:
```c
#include <windows.h>
#include <stdio.h>
#include <wchar.h>
#include "wcwidth.c"

int main(void)
{
    WCHAR ch = 0x0414;

    DWORD dw1 = GetTickCount();
    for (int i = 0; i < 100000000; ++i)
        WideCharToMultiByte(932, 0, &ch, 1, NULL, 0, NULL, NULL);
    DWORD dw2 = GetTickCount();
    for (int i = 0; i < 100000000; ++i)
        mk_wcwidth_cjk(ch);
    DWORD dw3 = GetTickCount();

    printf("%d, %d\n", dw2 - dw1, dw3 - dw2);
    return 0;
}
```
無題
`mk_wcwidth_cjk` is faster than `WideCharToMultiByte`.

File: win32ss/user/winsrv/consrv/frontends/terminal.c (hunk 1 above)
Comment (HBelusca) [reply win32ss/user/winsrv/consrv/frontends/terminal.c]: This was the idea, and in principle this concfg thing should be available in consrv already.

Comment (katahiromz) [reply win32ss/user/winsrv/consrv/frontends/terminal.c]: Compilation error.

Comment (HBelusca) [reply win32ss/user/winsrv/consrv/frontends/terminal.c]: Then the frontends/terminal.c file just misses a `#include "concfg/font.h"` include.

File: win32ss/user/winsrv/consrv/frontends/wcwidth.c (hunk 3 above)
Comment (HBelusca) [reply win32ss/user/winsrv/consrv/frontends/wcwidth.c]: Is this function to do the same job as this one? :
https://github.com/PowerShell/PSReadLine/blob/d648aed6da66e2c3947e71825fbc8e7009a30c3d/PSReadLine/Render.cs#L870

File: win32ss/user/winsrv/consrv/frontends/terminal.c (hunk 1 above)
Comment (katahiromz) [reply win32ss/user/winsrv/consrv/frontends/terminal.c]: OK.

File: win32ss/user/winsrv/consrv/frontends/wcwidth.c (hunk 3 above)
Comment (katahiromz) [reply win32ss/user/winsrv/consrv/frontends/wcwidth.c]: Yep, that is referring http://www.cl.cam.ac.uk/~mgk25/c/wcwidth.c .

Comment (katahiromz) [reply win32ss/user/winsrv/consrv/frontends/wcwidth.c]: That's MIT License. Can we borrow this code?

Comment (HBelusca) [reply win32ss/user/winsrv/consrv/frontends/wcwidth.c]: By rewriting it, possibly (in any case the logic should be equivalent...)

Comment (katahiromz) [reply win32ss/user/winsrv/consrv/frontends/wcwidth.c]: OK.

File: win32ss/user/winsrv/consrv/frontends/terminal.c (hunk 1 above)
Comment (katahiromz) [reply win32ss/user/winsrv/consrv/frontends/terminal.c]: No. Д (`U+0414`) must be a fullwidth character. PowerShell is wrong. PowerShell's bug:
無題

File: win32ss/user/winsrv/consrv/frontends/wcwidth.c (hunk 3 above)
Comment (katahiromz) [reply win32ss/user/winsrv/consrv/frontends/wcwidth.c]: No. Д (`U+0414`) must be a fullwidth character in CJK. PowerShell is wrong.
無題

Comment (katahiromz) [reply win32ss/user/winsrv/consrv/frontends/wcwidth.c]: Eventually I found PowerShell's bug...

Comment (katahiromz) [reply win32ss/user/winsrv/consrv/frontends/wcwidth.c]: Reverted.

Comment (katahiromz) [reply win32ss/user/winsrv/consrv/frontends/wcwidth.c]: Ummm...
無題
Д (U+0414) is a fullwidth character in Japanese. But it seems a halfwidth character in Chinese.

Comment (katahiromz) [reply win32ss/user/winsrv/consrv/frontends/wcwidth.c]: 無題
Even in Chinese, `U+0404` is a fullwidth... ???

Comment (katahiromz) [reply win32ss/user/winsrv/consrv/frontends/wcwidth.c]: Unifont is not standard Chinese font... I will follow SimSun.

**Comment by katahiromz** [comment]
OK. Ready to go.

**Pull request merged**